warehouse/upload_stats_test.go

```diff
@@ -0,0 +1,404 @@
+package warehouse
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/warehouse/internal/repo"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats/mock_stats"
+	migrator "github.com/rudderlabs/rudder-server/services/sql-migrator"
+
+	"github.com/golang/mock/gomock"
+
+	sqlmiddleware "github.com/rudderlabs/rudder-server/warehouse/integrations/middleware/sqlquerywrapper"
+
+	"github.com/ory/dockertest/v3"
+
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource"
+	"github.com/rudderlabs/rudder-server/warehouse/internal/model"
+)
+
+func TestUploadJob_Stats(t *testing.T) {
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+
+	pgResource, err := resource.SetupPostgres(pool, t)
+	require.NoError(t, err)
+
+	t.Log("db:", pgResource.DBDsn)
+
+	err = (&migrator.Migrator{
+		Handle:          pgResource.DB,
+		MigrationsTable: "wh_schema_migrations",
+	}).Migrate("warehouse")
+	require.NoError(t, err)
+
+	sqlStatement, err := os.ReadFile("testdata/sql/stats_test.sql")
+	require.NoError(t, err)
+
+	_, err = pgResource.DB.Exec(string(sqlStatement))
+	require.NoError(t, err)
```

Comment:
We could remove some duplicated code in these tests by using a common setup function:

```go
func setupUploadTest(t testing.TB, migrationsPath string) *sql.DB {
	pool, err := dockertest.NewPool("")
	require.NoError(t, err)

	pgResource, err := resource.SetupPostgres(pool, t)
	require.NoError(t, err)

	t.Log("db:", pgResource.DBDsn)

	err = (&migrator.Migrator{
		Handle:          pgResource.DB,
		MigrationsTable: "wh_schema_migrations",
	}).Migrate("warehouse")
	require.NoError(t, err)

	sqlStatement, err := os.ReadFile(migrationsPath)
	require.NoError(t, err)

	_, err = pgResource.DB.Exec(string(sqlStatement))
	require.NoError(t, err)

	return pgResource.DB
}
```